File: python/src/elap_ai/handler_examples.py
"""
Ejemplos de handlers REST usando el framework de error handling.

Estos ejemplos muestran los patrones recomendados para:
1. Validación de entrada
2. Verificación de recursos
3. Manejo de errores externos
4. Operaciones asíncronas con timeout
5. Respuestas HTTP estándar
"""

import asyncio
from aiohttp import web
from typing import Dict, Any, Optional

from .api_response import (
    handle_api_errors,
    success_response,
    created_response,
    accepted_response,
    error_response,
)
from .errors import (
    ValidationError,
    AgentNotFound,
    AgentTimeout,
    AgentExecutionFailed,
    DatabaseError,
    ElapException,
)


# ============================================================================
# === Ejemplo 1: Validación de entrada
# ============================================================================

@handle_api_errors()
async def crear_agente_handler(request: web.Request) -> web.Response:
    """
    POST /api/agents

    Patrón: Validación de entrada

    Antes:
        async def crear_agente(request):
            data = await request.json()
            try:
                agente = crear_agente_internal(data)
                return web.json_response({"agente": agente}, status=201)
            except Exception as e:
                return web.json_response(
                    {"error": str(e)},
                    status=400
                )

    Después:
        @handle_api_errors()
        async def crear_agente(request):
            data = await request.json()

            # ✅ Validar
            if not data.get('nombre'):
                raise ValidationError("Name is required")
            if not data.get('rol'):
                raise ValidationError("Role is required")
            if len(data['nombre']) > 256:
                raise ValidationError("Name too long (max 256 chars)")

            # ✅ Crear
            agente = crear_agente_internal(data)

            # ✅ Retornar
            return created_response(data=agente)
    """
    data = await request.json()

    # ✅ Validar entrada
    if not data.get('nombre'):
        raise ValidationError("Name is required")
    if not data.get('rol'):
        raise ValidationError("Role is required")
    if len(data['nombre']) > 256:
        raise ValidationError("Name too long (max 256 chars)")

    # ✅ Crear agente (simulado)
    agente = {
        'id': 'agent_123',
        'nombre': data['nombre'],
        'rol': data['rol'],
        'estado': 'creado',
    }

    # ✅ Retornar respuesta
    return created_response(
        data=agente,
        message=f"Agent {data['nombre']} created successfully"
    )


# ============================================================================
# === Ejemplo 2: Verificar recurso existe
# ============================================================================

@handle_api_errors()
async def get_agente_handler(request: web.Request) -> web.Response:
    """
    GET /api/agents/{agent_id}

    Patrón: Verificar que recurso existe

    Antes:
        async def get_agente(request):
            agent_id = request.match_info['agent_id']
            agente = repo.get(agent_id)
            if not agente:
                return web.json_response({"error": "Not found"}, status=404)
            return web.json_response(agente)

    Después:
        @handle_api_errors()
        async def get_agente(request):
            agent_id = request.match_info['agent_id']

            # ✅ Obtener
            agente = repo.get(agent_id)
            if not agente:
                raise AgentNotFound(f"Agent {agent_id} not found")

            # ✅ Retornar
            return success_response(data=agente)
    """
    agent_id = request.match_info.get('agent_id')

    # ✅ Validar parámetro
    if not agent_id:
        raise ValidationError("Agent ID is required")

    # ✅ Obtener recurso (simulado)
    if agent_id == 'valid_agent':
        agente = {
            'id': agent_id,
            'nombre': 'Test Agent',
            'rol': 'sales',
            'estado': 'activo',
        }
    else:
        raise AgentNotFound(f"Agent {agent_id} not found")

    # ✅ Retornar respuesta
    return success_response(data=agente)


# ============================================================================
# === Ejemplo 3: Ejecutar operación con timeout
# ============================================================================

@handle_api_errors()
async def ejecutar_agente_handler(request: web.Request) -> web.Response:
    """
    POST /api/agents/{agent_id}/execute

    Patrón: Ejecutar operación con timeout

    Antes:
        async def ejecutar_agente(request):
            try:
                data = await request.json()
                agent_id = request.match_info['agent_id']
                result = await agent.execute(data['prompt'])
                return web.json_response({"resultado": result})
            except asyncio.TimeoutError:
                return web.json_response({"error": "timeout"}, status=504)
            except Exception as e:
                return web.json_response({"error": str(e)}, status=500)

    Después:
        @handle_api_errors()
        async def ejecutar_agente(request):
            data = await request.json()
            agent_id = request.match_info['agent_id']

            # ✅ Validar
            if not data.get('prompt'):
                raise ValidationError("Prompt is required")

            # ✅ Obtener agente
            agente = repo.get(agent_id)
            if not agente:
                raise AgentNotFound(f"Agent {agent_id} not found")

            # ✅ Ejecutar con timeout
            try:
                timeout = data.get('timeout_ms', 30000) / 1000
                result = await asyncio.wait_for(
                    agente.execute(data['prompt']),
                    timeout=timeout
                )
            except asyncio.TimeoutError:
                raise AgentTimeout(f"Execution timeout after {timeout}s")
            except Exception as e:
                raise AgentExecutionFailed(str(e))

            # ✅ Retornar respuesta
            return success_response(data={
                'agent_id': agent_id,
                'resultado': result,
                'duracion_ms': 1500,
            })
    """
    data = await request.json()
    agent_id = request.match_info.get('agent_id', 'default')

    # ✅ Validar entrada
    if not data.get('prompt'):
        raise ValidationError("Prompt is required")

    # ✅ Validar timeout
    timeout_ms = data.get('timeout_ms', 30000)
    if timeout_ms < 1000:
        raise ValidationError("Timeout must be at least 1000ms")

    # ✅ Obtener agente (simulado)
    if agent_id not in ['hr', 'finance']:
        raise AgentNotFound(f"Agent {agent_id} not found")

    # ✅ Ejecutar con timeout
    try:
        timeout = timeout_ms / 1000
        # Simular ejecución del agente
        await asyncio.sleep(0.1)
        resultado = f"Resultado para: {data['prompt'][:50]}..."
    except asyncio.TimeoutError:
        raise AgentTimeout(f"Execution timeout after {timeout}s")
    except Exception as e:
        raise AgentExecutionFailed(f"Agent execution failed: {e}")

    # ✅ Retornar respuesta
    return success_response(
        data={
            'agent_id': agent_id,
            'resultado': resultado,
            'duracion_ms': 150,
        },
        message="Execution completed successfully"
    )


# ============================================================================
# === Ejemplo 4: Consultar base de datos
# ============================================================================

@handle_api_errors()
async def consultar_base_datos_handler(request: web.Request) -> web.Response:
    """
    GET /api/database/query

    Patrón: Convertir errores de base de datos
    """
    query = request.rel_url.query.get('q', 'SELECT * FROM agents')

    # ✅ Validar query
    if not query:
        raise ValidationError("Query parameter 'q' is required")
    if len(query) > 1000:
        raise ValidationError("Query too long (max 1000 chars)")

    # ✅ Ejecutar query (simulado)
    try:
        # Simular error de BD
        if 'INVALID' in query:
            raise RuntimeError("SQL Syntax Error: INVALID keyword")

        resultados = [
            {'id': '1', 'nombre': 'Agent 1'},
            {'id': '2', 'nombre': 'Agent 2'},
        ]
    except RuntimeError as e:
        raise DatabaseError(f"Database query failed: {e}")

    # ✅ Retornar respuesta
    return success_response(
        data=resultados,
        message=f"Query executed: {len(resultados)} rows"
    )


# ============================================================================
# === Ejemplo 5: Operación asíncrona (202 Accepted)
# ============================================================================

@handle_api_errors()
async def generar_documento_handler(request: web.Request) -> web.Response:
    """
    POST /api/documents/generate

    Patrón: Operación asíncrona (enqueue, no esperar)

    Respuesta: 202 Accepted (en lugar de 201/200)
    """
    data = await request.json()

    # ✅ Validar entrada
    if not data.get('template_id'):
        raise ValidationError("Template ID is required")
    if not data.get('employee_name'):
        raise ValidationError("Employee name is required")

    # ✅ Enqueue job (no esperar)
    # En una aplicación real, esto iría a una cola de trabajos
    job_id = f"job_{hash(str(data)) % 10000}"

    # ✅ Retornar 202 Accepted
    return accepted_response(
        data={
            'job_id': job_id,
            'status': 'queued',
            'template_id': data['template_id'],
        },
        message="Document generation queued"
    )


# ============================================================================
# === Resumen de patrones
# ============================================================================

"""
TABLA DE PATRONES EN PYTHON:

┌──────────────────┬────────────────────────────────────┬───────────┐
│ Situación        │ Patrón                             │ Error     │
├──────────────────┼────────────────────────────────────┼───────────┤
│ Entrada inválida │ raise ValidationError("...")       │ 400       │
│ Recurso existe   │ raise AgentNotFound("...")         │ 404       │
│ Sin permiso      │ raise PermissionDenied("...")      │ 403       │
│ Timeout          │ raise AgentTimeout("...")          │ 504       │
│ Ejecución falla  │ raise AgentExecutionFailed("...")  │ 500       │
│ Error de BD      │ raise DatabaseError("...")         │ 500       │
│ Operación async  │ return accepted_response(...)      │ 202       │
└──────────────────┴────────────────────────────────────┴───────────┘

DECORADOR @handle_api_errors():

✅ Captura ElapException automáticamente
✅ Convierte a respuesta HTTP JSON
✅ Incluye status codes correctos
✅ Captura excepciones desconocidas
✅ Mensajes seguros para usuario

ESTRUCTURA DE HANDLER:

@handle_api_errors()
async def handler(request: web.Request) -> web.Response:
    # 1. Extraer datos
    data = await request.json()
    id = request.match_info.get('id')

    # 2. Validar entrada
    if not data.get('campo'):
        raise ValidationError("Field is required")

    # 3. Verificar recurso
    recurso = repo.get(id)
    if not recurso:
        raise AgentNotFound(f"Resource {id} not found")

    # 4. Ejecutar lógica
    try:
        resultado = await operacion(data)
    except TimeoutError:
        raise AgentTimeout("Operation took too long")
    except Exception as e:
        raise AgentExecutionFailed(str(e))

    # 5. Retornar respuesta
    return success_response(data=resultado)
"""

File: crates/elap-core/src/api/handler_examples.rs
/// Ejemplos de handlers usando el nuevo framework de error handling.
///
/// Estos ejemplos muestran los patrones recomendados para:
/// 1. Validación de entrada
/// 2. Verificación de recursos
/// 3. Control de acceso (RBAC)
/// 4. Conversión de errores externos
/// 5. Propagación de errores con contexto

use axum::{
    extract::{Path, State},
    Json,
};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::error::ElapError;
use super::error_response::{ApiResult, SuccessResponse};

// ============================================================================
// === Estructuras de solicitud/respuesta
// ============================================================================

#[derive(Deserialize)]
pub struct CrearAgentRequest {
    pub nombre: String,
    pub rol: String,
    pub objetivo: String,
}

#[derive(Serialize)]
pub struct AgentResponse {
    pub id: String,
    pub nombre: String,
    pub rol: String,
    pub objetivo: String,
}

#[derive(Deserialize)]
pub struct EjecutarAgentRequest {
    pub entrada: String,
    pub timeout_ms: Option<u64>,
}

#[derive(Serialize)]
pub struct EjecucionResponse {
    pub id: String,
    pub resultado: String,
    pub duracion_ms: u64,
}

// ============================================================================
// === Ejemplo 1: Validación de entrada
// ============================================================================

/// Patrón: Validar datos de entrada
///
/// Antes:
/// ```ignore
/// pub async fn crear_agente(Json(payload): Json<CrearAgentRequest>) -> ... {
///     let agente = crear_agente_internal(&payload.nombre);
/// }
/// ```
///
/// Después:
/// ```ignore
/// pub async fn crear_agente(Json(payload): Json<CrearAgentRequest>) -> ApiResult<Json<SuccessResponse<AgentResponse>>> {
///     // Validar
///     if payload.nombre.is_empty() {
///         return Err(ElapError::ValidationError("Name cannot be empty".to_string()));
///     }
///     if payload.nombre.len() > 256 {
///         return Err(ElapError::ValidationError("Name too long (max 256 chars)".to_string()));
///     }
///     if !["sales", "hr", "finance"].contains(&payload.rol.as_str()) {
///         return Err(ElapError::ValidationError(format!("Invalid role: {}", payload.rol)));
///     }
///
///     // Crear
///     let agente = crear_agente_internal(&payload);
///     Ok(Json(SuccessResponse::created(agente)))
/// }
/// ```

pub async fn ejemplo_validacion(
    Json(payload): Json<CrearAgentRequest>,
) -> ApiResult<Json<SuccessResponse<AgentResponse>>> {
    // ✅ Validar entrada
    payload.nombre.is_empty().then(|| {
        Err(ElapError::ValidationError(
            "Name cannot be empty".to_string(),
        ))
    });

    payload.rol.is_empty().then(|| {
        Err(ElapError::ValidationError(
            "Role cannot be empty".to_string(),
        ))
    });

    // ✅ Crear recurso
    let agente = AgentResponse {
        id: Uuid::new_v4().to_string(),
        nombre: payload.nombre,
        rol: payload.rol,
        objetivo: payload.objetivo,
    };

    Ok(Json(SuccessResponse::created(agente)))
}

// ============================================================================
// === Ejemplo 2: Verificar recurso existe
// ============================================================================

/// Patrón: Verificar que un recurso existe antes de operarlo
///
/// Antes:
/// ```ignore
/// pub async fn get_agent(Path(id): Path<String>) -> ... {
///     match repo.get(&id) {
///         Some(agent) => (StatusCode::OK, Json(agent)),
///         None => (StatusCode::NOT_FOUND, Json(ErrorResponse { ... })),
///     }
/// }
/// ```
///
/// Después:
/// ```ignore
/// pub async fn get_agent(Path(id): Path<String>) -> ApiResult<Json<AgentResponse>> {
///     let agent = REPO.lock()
///         .get(&id)
///         .ok_or_else(|| ElapError::AgentNotFound(
///             format!("Agent {} not found", id)
///         ))?;
///
///     Ok(Json(agent))
/// }
/// ```

pub async fn ejemplo_get_agent(
    Path(id): Path<String>,
) -> ApiResult<Json<AgentResponse>> {
    // Simulado: repo.get(&id)
    let _agent_exists = id == "valid_id";

    if !_agent_exists {
        return Err(ElapError::AgentNotFound(format!(
            "Agent {} not found",
            id
        )));
    }

    // ✅ Retornar recurso
    Ok(Json(AgentResponse {
        id,
        nombre: "Test Agent".to_string(),
        rol: "sales".to_string(),
        objetivo: "test".to_string(),
    }))
}

// ============================================================================
// === Ejemplo 3: Control de acceso (RBAC)
// ============================================================================

/// Patrón: Validar permisos antes de ejecutar acción
///
/// ```ignore
/// pub async fn delete_agent(
///     claims: Claims,  // JWT del usuario
///     Path(id): Path<String>,
/// ) -> ApiResult<Json<Value>> {
///     // ✅ Validar autenticación
///     if claims.user_id.is_empty() {
///         return Err(ElapError::Unauthorized(
///             "User not authenticated".to_string()
///         ));
///     }
///
///     // ✅ Validar permiso (RBAC)
///     let rbac = ValidadorRBAC::new();
///     rbac.validar(claims.role, Accion::Eliminar, "agent")
///         .map_err(|e| ElapError::PermissionDenied(
///             format!("User {} lacks permission: {}", claims.user_id, e)
///         ))?;
///
///     // ✅ Ejecutar
///     REPO.lock().remove(&id)?;
///
///     Ok(Json(json!({"status": "deleted"})))
/// }
/// ```

// ============================================================================
// === Ejemplo 4: Convertir errores externos
// ============================================================================

/// Patrón: Convertir errores de librerías externas
///
/// ```ignore
/// pub async fn query_database() -> ApiResult<Vec<String>> {
///     let results = db.query("SELECT * FROM agents")
///         .await
///         .map_err(|e| ElapError::DatabaseError(
///             format!("Query failed: {}", e)
///         ))?;
///
///     Ok(results)
/// }
/// ```

pub async fn ejemplo_error_conversion() -> ApiResult<Vec<String>> {
    // Simulado: db.query()
    let result: Result<Vec<String>, String> =
        Err("Connection timeout".to_string());

    // ✅ Convertir error externo a ElapError
    let results = result.map_err(|e| {
        ElapError::DatabaseError(format!("Database query failed: {}", e))
    })?;

    Ok(results)
}

// ============================================================================
// === Ejemplo 5: Error con contexto
// ============================================================================

/// Patrón: Agregar contexto (user_id, request_id, etc.) a errores
///
/// ```ignore
/// pub async fn execute_agent(
///     claims: Claims,
///     Path(id): Path<String>,
///     Json(payload): Json<EjecutarAgentRequest>,
/// ) -> ApiResult<Json<EjecucionResponse>> {
///     // ✅ Obtener agente
///     let agent = REPO.lock()
///         .get(&id)
///         .ok_or_else(|| ElapError::AgentNotFound(
///             format!("Agent {} not found", id)
///         ))?;
///
///     // ✅ Ejecutar con timeout
///     let timeout = Duration::from_millis(payload.timeout_ms.unwrap_or(30000));
///     let result = tokio::time::timeout(timeout, agent.execute(&payload.entrada))
///         .await
///         .map_err(|_| {
///             ElapError::AgentTimeout(
///                 format!("Execution timeout after {:?}", timeout)
///             )
///         })?
///         .map_err(|e| {
///             ElapError::AgentExecutionFailed(
///                 format!("Agent execution failed: {}", e)
///             )
///         })?;
///
///     Ok(Json(SuccessResponse::ok(EjecucionResponse {
///         id: Uuid::new_v4().to_string(),
///         resultado: result,
///         duracion_ms: 1500,
///     })))
/// }
/// ```

// ============================================================================
// === Ejemplo 6: Resultado con recuperación automática
// ============================================================================

/// Patrón: Retry automático en errores recuperables
///
/// ```ignore
/// async fn ejecutar_con_retry<F, T>(
///     mut f: F,
///     max_intentos: u32,
/// ) -> ApiResult<T>
/// where
///     F: FnMut() -> BoxFuture<'static, ApiResult<T>>,
/// {
///     for intento in 1..=max_intentos {
///         match f().await {
///             Ok(result) => return Ok(result),
///             Err(e) => {
///                 // Retry solo en errores recuperables
///                 if !e.es_recuperable() {
///                     return Err(e);
///                 }
///                 if intento == max_intentos {
///                     return Err(ElapError::InternalError(
///                         format!("Failed after {} retries", max_intentos)
///                     ));
///                 }
///                 tokio::time::sleep(Duration::from_millis(100 * intento as u64)).await;
///             }
///         }
///     }
///     unreachable!()
/// }
/// ```

// ============================================================================
// === Resumen de patrones
// ============================================================================

/*
TABLA DE PATRONES:

┌─────────────────────────┬──────────────────────────────────┬────────────────┐
│ Situación               │ Patrón                           │ Error a usar   │
├─────────────────────────┼──────────────────────────────────┼────────────────┤
│ Entrada inválida        │ Validar → ValidationError        │ 400            │
│ Recurso no existe       │ Check → NotFound/AgentNotFound   │ 404            │
│ Recurso existe ya       │ Check → Conflict                 │ 409            │
│ Sin autenticación       │ Check → Unauthorized             │ 401            │
│ Sin permiso (RBAC)      │ Validar → PermissionDenied       │ 403            │
│ DB query falla          │ map_err → DatabaseError          │ 500            │
│ gRPC timeout            │ timeout → GrpcTimeout/AgentTimeout│ 504            │
│ Tool no encontrada      │ Check → ToolNotFound             │ 404            │
│ Tool execution falla    │ map_err → ToolExecutionFailed    │ 500            │
│ Pool exhausto           │ Check → ConnectionPoolExhausted  │ 503            │
│ Servicio no disponible  │ Check → ConnectionUnavailable    │ 503            │
│ Error externo           │ map_err → wrap → ElapError       │ 500            │
│ No implementado         │ Check → NotImplemented           │ 501            │
└─────────────────────────┴──────────────────────────────────┴────────────────┘

ESTRUCTURA DE HANDLER:

pub async fn handler(
    // ✅ 1. Extraer datos
    Path(id): Path<String>,
    Json(payload): Json<Request>,
    // ✅ 2. Autenticación (si aplica)
    claims: Claims,
) -> ApiResult<Json<SuccessResponse<Response>>> {
    // ✅ 3. Validar entrada
    if payload.is_invalid() {
        return Err(ElapError::ValidationError("...".to_string()));
    }

    // ✅ 4. Verificar permisos
    validador_rbac.validar(...)?;

    // ✅ 5. Obtener recurso
    let resource = repo.get(&id)
        .ok_or(ElapError::NotFound(...))?;

    // ✅ 6. Ejecutar lógica
    let result = operation()
        .await
        .map_err(|e| ElapError::InternalError(...))?;

    // ✅ 7. Retornar respuesta
    Ok(Json(SuccessResponse::ok(result)))
}
*/
